rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+
+- each Jump Opcode jumps to a valid location in the bytecode.
+- each constant references a valid index in the constants list
+- each path index references a valid path.
+
+It should be possible to run the linter after every compile (at Vector boot
+time), so the linter should run relatively fast.
+
+### Testing
+
+The VM will need thorough testing since it is possible to generate invalid
+bytecode. For example, the following bytecode is invalid:
+
+```rust
+OpCode(CONSTANT)
+Primitive(1)
+OpCode(ADD)
+```
+
+The `ADD` OpCode expects there to be two values on the stack to add
+together. We should never get into the situation where the VRL compiler can
+generate this.
+
+With property testing we can generate a vast combination of valid VRL ASTs,
+compile this to bytecode and then run the VM to ensure we get no buffer
+underflows, invalid constant locations or other invalid bytecode.
+
+Given that implementing the VM involves adding methods to the Expression trait,
+it should be possible to run the existing AST alongside the VM. We can then
+ensure that we get exactly the same results from each approach.
+
+## Rationale
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
+
+## Drawbacks
+
+Downsides to using a Vm:
+
+- The code is a bit more complex. With an AST that is walked it is fairly
+  apparent what the code will be doing at any point. With a Vm, this is not the
+  case, it is harder to look at the instructions in the Vm and follow back to
+  what part of the VRL code is being evaluated. We will need to write some
+  extensive debugging tools to allow for decent introspection into the Vm.
+
+  However, VRL is a very simple language, which does also mean the VM will be
+  simple.
+
+- We lose some safety that we get from the Rust compiler. There will need
+  to be significant fuzz testing to ensure that the code runs correctly under
+  all circumstances.
+
+  It should be noted however that the entire VM will not require any _unsafe_
+  Rust code.
+
+- Currently each stdlib function is responsible for evaluating their own
+  parameters. This allows parameters to be lazily evaluated. With the Vm, the
+  parameters will need to be evaluated up front and the stack passed into the
+  function. This could impact performance.
+
+## Prior Art
+
+- GoScript - An implementation of Go using a bytecode Vm,
+  https://github.com/oxfeeefeee/goscript
+
+- CPython - https://github.com/python/cpython
+
+## Alternatives
+
+### WASM
+
+Instead of writing our own VM we could transpile VRL code to WASM. WASM is a
+mature and performant VM that could give us near native speeds.
+
+The issues with using WASM are that it would require data to be serialized in
+order to move between Vector and WASM. This would incur a significant
+performance penalty.
```

Comment:
I mentioned this in our meeting, but this isn't actually accurate. Wasm allows [both the host and the Wasm module to mutably access the same chunk of memory](https://webassembly.github.io/spec/core/exec/runtime.html#memory-instances).

See, for example, the [`wasmer`](https://docs.rs/wasmer/2.0.0/wasmer/struct.Memory.html) crate.

It's true, though, that we'll need wrappers to represent our types, but there's also a lot of ongoing work with [Wasm interface types](https://hacks.mozilla.org/2019/08/webassembly-interface-types/) to make this simpler.

Still, I agree with the conclusion that this is a step too far for a first iteration, but I want to avoid that conclusion being based on the wrong assumptions.